src/client/serialize/daos_serialize.c

```diff
@@ -1175,3 +1245,1833 @@ daos_cont_deserialize_attrs(char *filename, uint64_t *num_attrs,
+		      daos_handle_t *oh, hsize_t dkey_dims[], hsize_t akey_dims[],
+		      int *total_dkeys, int *total_akeys, uint64_t *bytes_written)
+{
+	int		rc = 0;
+	int		j = 0;
+	daos_key_t	diov;
+	char		dkey[ENUM_KEY_BUF] = {0};
+	uint64_t	ak_off = 0;
+	uint64_t	ak_next = 0;
+	uint64_t	total_akeys_this_dkey = 0;
+	int		k = 0;
+	hvl_t		*dkey_val;
+	hvl_t		*rec_kv_val;
+	daos_size_t	kv_single_size = 0;
+
+	for (j = 0; j < *total_dkeys_this_oid; j++) {
+		memset(&diov, 0, sizeof(diov));
+		memset(dkey, 0, sizeof(dkey));
+		dkey_val = &(hdf5->dkey_data)[*dk_off + j].dkey_val;
+		rec_kv_val = &(hdf5->dkey_data)[*dk_off + j].rec_kv_val;
+		memcpy(dkey, dkey_val->p, dkey_val->len);
+		d_iov_set(&diov, (void*)dkey_val->p, dkey_val->len);
+		ak_off = hdf5->dkey_data[*dk_off + j].akey_offset;
+		ak_next = 0;
+		total_akeys_this_dkey = 0;
+		if (*dk_off + j + 1 < (int)dkey_dims[0]) {
+			ak_next = hdf5->dkey_data[(*dk_off + j) + 1].akey_offset;
+			total_akeys_this_dkey = ak_next - ak_off;
+		} else if (*dk_off + j == ((int)dkey_dims[0] - 1)) {
+			total_akeys_this_dkey = ((int)akey_dims[0]) - ak_off;
+		}
+
+		/* if rec_kv_val.len != 0 then skip akey iteration, we can write data back
+		 * into DAOS using the daos_kv.h API using just oid, dkey, and key value
+		 * (stored in dkey dataset)
+		 */
+
+		/* run daos_kv_put on rec_kv_val (dkey val) and key (dkey) */
+		/* skip akey iteration for DAOS_OF_KV_FLAT objects
+		 */
+		if (rec_kv_val->len > 0) {
+			kv_single_size = rec_kv_val->len;
+			rc = daos_kv_put(*oh, DAOS_TX_NONE, 0, dkey, kv_single_size,
+					 rec_kv_val->p, NULL);
+			if (rc != 0) {
+				D_ERROR("failed to write kv object "DF_RC"\n", DP_RC(rc));
+				D_GOTO(out, rc);
+  			}
+			(*bytes_written) += kv_single_size;
+		} else {
+			for (k = 0; k < total_akeys_this_dkey; k++) {
+				/* TODO call cont_deserialize_akeys here */
+				rc = cont_deserialize_akeys(hdf5, diov, &ak_off, k, oh,
+							    total_akeys, bytes_written);
+				if (rc != 0) {
+					D_ERROR("failed to deserialize akeys "DF_RC"\n", DP_RC(rc));
+					D_GOTO(out, rc);
+				}
+			}
+		}
+		(*total_dkeys)++;
+	}
+out:
+	return rc;
+}
+
+int
+daos_cont_deserialize(int *total_oids, int *total_dkeys, int *total_akeys, uint64_t *bytes_written,
+		      daos_handle_t coh, char *filename)
+{
+	int			rc = 0;
+	int			i = 0;
+	hid_t			status = 0;
+	int			oid_ndims = 0;
+	int			dkey_ndims = 0;
+	int			akey_ndims = 0;
+	hsize_t			oid_dims[1];
+	hid_t			oid_dset = -1;
+	hid_t			oid_dspace = -1;
+	hid_t			oid_dtype = -1;
+	hsize_t			dkey_dims[1];
+	hid_t			dkey_dset = -1;
+	hid_t			dkey_dspace = -1;
+	hid_t			dkey_vtype = -1;
+	hsize_t			akey_dims[1];
+	hid_t			akey_dset = -1;
+	hid_t			akey_dspace = -1;
+	hid_t			akey_vtype = -1;
+	struct hdf5_args	hdf5 = {0};
+	daos_obj_id_t		oid;
+	daos_handle_t		oh;
+	bool			is_kv = false;
+	uint64_t                total_dkeys_this_oid = 0;
+    	uint64_t                dk_off = 0;
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs